### DISCUSSION TITLE
RigidBodyMotion behavior

### DISCUSSION THREAD

**Discussion by Marcraven** [FABLE-3DXRD/xrd_simulator#6] [General]
This is a continuation of a discussion which started via mail, but we consider it might be interesting for other users. I wanted to  perform a raster scan with the beam in the Y axis, then rotate the sample by one degree and then scan again. This whole process for 360 degrees. The script had the following structures:

```
  polycrystal = Polycrystal.load(os.path.join(dir_this_file,samples_directory ,f'{polycrystal_file}.pc'))
  beam = Beam.load(os.path.join(dir_this_file,beams_directory ,f'{beam_file}.beam'))
  detector = Detector.load(os.path.join(dir_this_file,detectors_directory ,f'{detector_file}.det'))
  
  
  origin_motion = RigidBodyMotion(rotation_axis=np.array([0, 1, 0]),
                           rotation_angle=np.radians(0.0001),
                           translation=np.array([0, -500, 0]))
  
  translate_motion = RigidBodyMotion(rotation_axis=np.array([0, 1, 0]),
                           rotation_angle=np.radians(0.0001),
                           translation=np.array([0, 1000, 0]))
  
  restart_translation = RigidBodyMotion(rotation_axis=np.array([0, 1, 0]),
                           rotation_angle=np.radians(0.0001),
                           translation=np.array([0, 500, 0]))
  
  rotate_motion = RigidBodyMotion(rotation_axis=np.array([0, 0, 1]),
                           rotation_angle=np.radians(1),
                           translation=np.array([0, 0, 0]))
  
  polycrystal.transform(origin_motion,time=1)
  for angle in range(361):

    
    polycrystal.diffract(beam=beam,
                    detector=detector,
                    rigid_body_motion=translate_motion,
                    number_of_processes=threads,
                    number_of_frames=67)

    diffraction_pattern = detector.render(frames_to_render='all',
                                    lorentz=True,
                                    polarization=False,
                                    structure_factor=True,
                                    method=method,#"centroid""centroid_with_scintillator",#"project",
                                    number_of_processes=threads,
                                    verbose=False)

    file_dir = os.path.join(tiffs_directory,f'angle_{angle}_pixel.tif')
    tifffile.imwrite(file_dir,diffraction_pattern.astype(np.float32), imagej=True)
    print(f'Finished projection number at {angle} degrees')
    del diffraction_pattern
    polycrystal.transform(restart_translation,time=1)
    polycrystal.transform(rotate_motion,time=1)
    polycrystal.transform(origin_motion,time=1)
```
The rationale behind all these _RigidBodyMotions_ is;

1) Before the for loop, we displace the sample 500um towards the right (_origin_motion_) to initialize the Y scan at the right spot.
2) _polycrystal.diffract_ then takes place over a 1000um translation to the left with 67 frames. This, I thought does not modify the actual rotation of the mesh since it takes place in diffract and not in transform.
3) A the end of the iteration, _restart_translation_ is applied to undo the initial 500um displacement, so the 1º degree rotation (_rotate_motion_) takes place around the (0,0,0) center of the sample
4) After rotation for 1º, the sample is back again applied _origin_motion_ so a new scan can start at the next iteration.

If  I understood correctly polycrystal.diffract actually leaves the sample transformed?

I also have the issue that the second iteration gives rise to 2*67 frames, for some reason it looks like the frames accumulate over iterations. ¿Is that possible?

Cheers,

PD: In polycrystal.diffract I now empty the detector.frames list before the frame creation and it seems to only then consider the current diffraction events instead of accumulating.

**Comment by AxelHenningsson** [comment]
Dear Marcraven,

Sorry for the confusion. In the current version of `xrd_simulator` I belive that motions only get applied when calling the `polycrystal.transform()`.

Diffraction events will accumulate like you say. A frame is simply a collection of scattering units that you may render whenever you like.  At any time you may empty  (`detector.frames = [] ` ) the detector or delete specific frames ( `del detector.frames[3:7]` ) if you fancy.
  
As far as I can see your script looks ok! What seems to be the problem? Have you tried to track the sample centroid with `polycrystal.mesh_lab.coord.mean(axis=0)`? Is it drifting?

Best,
Axel

**Comment by AxelHenningsson** [comment]
If you have the files I could try to debug your snippet locally if you like.

I am keen to make sure there is no issue in the repo. 

Best,
Axel

**Comment by Marcraven** [comment]
Hi @AxelHenningsson,

Is it possible that the rotation and translation don't take place homogeneously over all frames?

It feels like the first frame is very intense( maybe accumulating all the rotation), and the following frames have fewer spots in comparison.

**Comment by AxelHenningsson** [comment]
Hi @Marcraven 

I doubt this. it would be a major bug in `xrd_simulator` that would hopefully come up many times in the unit tests.

Looking at your code again I see that you have a very small rotation for the diffraction motion (`translate_motion`).

A pure translation of a rigid polycrystal will not change the angle of incidence between the Miller planes and the direct beam. Thus a pure translation will not give rise to diffraction unless the model was to include wavelength broadening or continuous lattice orientation fields.  `xrd_simulator`  simulates a polycrystal as composed of a finite set of single (perfect) crystals (each is a tetrahedron in the mesh) and considers the wavelength fixed. It is therefore not meaningful to read out frames without including some rotational interval. If you try and set the rotation angle to 0.1 or 0.2 degrees in your diffraction motion I believe you will see a dramatic difference in your patterns. 

Cheers
Axel

**Comment by AxelHenningsson** [comment]
@Marcraven : Is there still some doubt related to the output of your simulation? In this case I suggest that you provide access to a full reproducible example.  Otherwise, If there is nothing else I may assist you with in relation to this discussion (which I am happy to do), I will close this thread as completed. 

Best,
Axel
